modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+    updateBasisMolalityForKnownActivity(_basis_molality);
+    computeRemainingBasisActivities(_basis_activity);
+
+    // Step 2: compute equilibrium molality based on the activities just computed
+    computeEqmMolalities(_eqm_molality);
+  }
+
+  computeBulk(_bulk_moles);
+  computeFreeMineralMoles(_basis_molality);
+  computeSorbingSurfaceArea(_sorbing_surface_area);
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getChargeBalanceBasisIndex() const
+{
+  return _charge_balance_basis_index;
+}
+
+Real
+EquilibriumGeochemicalSystem::getLog10K(unsigned j) const
+{
+  if (j >= _num_eqm)
+    mooseError("Cannot retrieve log10K for equilibrium species ",
+               j,
+               " since there are only ",
+               _num_eqm,
+               " equilibrium species");
+  return _eqm_log10K[j];
+}
+
+unsigned
+EquilibriumGeochemicalSystem::getNumRedox() const
+{
+  return _num_redox;
+}
+
+Real
+EquilibriumGeochemicalSystem::getRedoxLog10K(unsigned red) const
+{
+  if (red >= _num_redox)
+    mooseError("Cannot retrieve log10K for redox species ",
+               red,
+               " since there are only ",
+               _num_redox,
+               " redox species");
+  return _redox_log10K[red];
+}
+
+Real
+EquilibriumGeochemicalSystem::log10RedoxActivityProduct(unsigned red) const
+{
+  if (red >= _num_redox)
+    mooseError("Cannot retrieve activity product for redox species ",
+               red,
+               " since there are only ",
+               _num_redox,
+               " redox species");
+  Real log10ap = 0.0;
+  for (unsigned basis_i = 0; basis_i < _num_basis; ++basis_i)
+    log10ap += _mgd.redox_stoichiometry(red, basis_i) * std::log10(_basis_activity[basis_i]);
+  return log10ap;
+}
+
+void EquilibriumGeochemicalSystem::buildTemperatureDependentQuantities(Real /*temperature*/)
+{
+  // When temperature least-squares stuff has been done, this function will be altered
+  // at 25degC currently
+  for (unsigned eqm_j = 0; eqm_j < _num_eqm; ++eqm_j)
+    _eqm_log10K[eqm_j] = _mgd.eqm_log10K(eqm_j, 1);
+  for (unsigned red = 0; red < _num_redox; ++red)
+    _redox_log10K[red] = _mgd.redox_log10K(red, 1);
+}
+
+void
+EquilibriumGeochemicalSystem::buildAlgebraicInfo(std::vector<bool> & in_algebraic_system,
+                                                 unsigned & num_basis_in_algebraic_system,
+                                                 unsigned & num_in_algebraic_system,
+                                                 std::vector<unsigned> & algebraic_index,
+                                                 std::vector<unsigned> & basis_index) const
+{
+  // build in_algebraic_system
+  for (const auto & name_index : _mgd.basis_species_index)
+  {
+    const std::string name = name_index.first;
+    const unsigned basis_ind = name_index.second;
+    const ConstraintMeaningEnum meaning = _constraint_meaning[basis_ind];
+    if (name == "H2O")
```

Comment:
Same here. Do we have a 'H2O_index' to use instead?

Reply:
Yea, this is just in the constructor, so probably OK.